### PR TITLE
Buoyancy perturbation mode

### PR DESCRIPTION
**Pull request by matt-frey**

### CODE REVIEW

File: src/3d/parcels/parcel_interpl.f90 [hunk 1]
```diff
@@ -170,6 +173,11 @@ subroutine par2grid(l_reuse)
 #else
                 btot = parcels%buoyancy(n)
 #endif
+
+#ifdef ENABLE_BUOYANCY_PERTURBATION_MODE
+                ! remove basic state N^2 * z
+                btot = btot - bfsq * parcels%position(3, n)
```
Comment (daviddritschel) [inline src/3d/parcels/parcel_interpl.f90:179]: So this defines b' on parcels, and is this then fed into par2grid to compute the gridded field?

File: src/3d/stepper/rk_utils.f90 [hunk 2]
```diff
@@ -166,6 +169,10 @@ function get_time_step(t) result(dt)
             gradb = f12 * dxi(I_Z) * (tbuoyg(1:nz+1, box%lo(2):box%hi(2), box%lo(1):box%hi(1)) &
                                     - tbuoyg(-1:nz-1, box%lo(2):box%hi(2), box%lo(1):box%hi(1)))
 
+#ifdef ENABLE_BUOYANCY_PERTURBATION_MODE
+            gradb = bradb + bfsq
```
Comment (daviddritschel) [inline src/3d/stepper/rk_utils.f90:173]: what is bradb on the rhs?  Is this supposed to be gradb?

File: src/3d/fields/field_netcdf.f90 [hunk 3]
```diff
@@ -242,6 +258,15 @@ subroutine write_netcdf_fields(t)
 #endif
             call write_field_double(NC_VOL, volg, start, cnt)
 
+#ifdef ENABLE_BUOYANCY_PERTURBATION_MODE
+            ! remove basic state
+            do iz = 0, nz
+                z = lower(3) + dble(iz) * dx(3)
+                tbuoyg(iz, :, :) = tbuoyg(iz, :, :) - bfsq * z
```
Comment (daviddritschel) [inline src/3d/fields/field_netcdf.f90:265]: Why should tbuoyg contain the total gridded buoyancy here?  I thought the gridded buoyancy found from par2grid is actually only the perturbation buoyancy.

File: src/3d/fields/field_netcdf.f90 [hunk 4]
```diff
@@ -231,6 +239,14 @@ subroutine write_netcdf_fields(t)
             call write_field_double(NC_Y_VOR, vortg(:, :, :, 2), start, cnt)
             call write_field_double(NC_Z_VOR, vortg(:, :, :, 3), start, cnt)
 
+#ifdef ENABLE_BUOYANCY_PERTURBATION_MODE
+            ! add basic state
+            do iz = 0, nz
+                z = lower(3) + dble(iz) * dx(3)
+                tbuoyg(iz, :, :) = tbuoyg(iz, :, :) + bfsq * z
```
Comment (daviddritschel) [inline src/3d/fields/field_netcdf.f90:246]: This is presumably only for output so the entire buoyancy field is written?

**Review by daviddritschel — COMMENTED** [review]
Please see my comments - one might be serious (bradb one)

**Review by daviddritschel — APPROVED** [review]

File: src/3d/parcels/parcel_interpl.f90 (hunk 1 above)
Comment (matt-frey) [reply src/3d/parcels/parcel_interpl.f90:179]: Yes.

File: src/3d/stepper/rk_utils.f90 (hunk 2 above)
Comment (matt-frey) [reply src/3d/stepper/rk_utils.f90]: Indeed, I realised when I wanted to compile the code.

File: src/3d/fields/field_netcdf.f90 (hunk 3 above)
Comment (matt-frey) [reply src/3d/fields/field_netcdf.f90:265]: This is for writing the field.

**Pull request merged**